sdk/src/OpenTelemetry/Processor/Batch.hs

```diff
@@ -213,20 +227,37 @@ data ProcessorMessage = Flush | Shutdown
 batchProcessor :: MonadIO m => BatchTimeoutConfig -> Exporter ImmutableSpan -> m Processor
 batchProcessor BatchTimeoutConfig {..} exporter = liftIO $ do
   batch <- newIORef $ boundedMap maxQueueSize
-  workSignal <- newEmptyMVar
-  worker <- async $ loop $ do
-    req <-
-      liftIO $
-        timeout (millisToMicros scheduledDelayMillis) $
-          takeMVar workSignal
-    batchToProcess <- liftIO $ atomicModifyIORef' batch buildExport
-    res <- liftIO $ Exporter.exporterExport exporter batchToProcess
-
-    -- if we were asked to shutdown, quit cleanly after this batch
-    -- FIXME: this could lose batches if there's more than one in queue?
-    case req of
-      Just Shutdown -> throwE res
-      _ -> pure ()
+  workSignal <- newEmptyTMVarIO
+  shutdownSignal <- newEmptyTMVarIO
+  let workerAction = do
+        delay <- registerDelay (millisToMicros scheduledDelayMillis)
+        req <- atomically $ do
+          msum
+            [ const Flush <$> do
+                continue <- readTVar delay
+                check continue
+            , const Flush <$> takeTMVar workSignal
+            , const Shutdown <$> takeTMVar shutdownSignal
+            ]
+        batchToProcess <- atomicModifyIORef' batch buildExport
+        -- we mask async exceptions in this, so that a buggy exporter that
+        -- catches async exceptions won't swallow them. since we use
+        -- an interruptible mask, blocking calls can still be killed, like
+        -- `threadDelay` or `putMVar` or most file I/O operations.
+        --
+        -- if we've received a shutdown, then we should be expecting
+        -- a `cancel` anytime now.
+        res <- mask_ $ Exporter.exporterExport exporter batchToProcess
```

Comment:
i'm not 100% happy with this, but the ability of any `Exporter` being able to `try x :: IO (Either SomeException a)` and catch async exceptions makes this pretty tricky.

Realistically, any exporter is going to be doing `threadDelay` or some other kind of blocking IO, and it'll receive the async exception anyway. `mask_` or not probably doesn't matter. Eg: `try $ threadDelay 10000` will still catch the async exception.

If we wanted to be *really* sure about the action not swallowing async exceptions, we'd need to put it in an`uninterruptibleMask`. *IF* the action completes, then the async exception would be received after the `uninterruptibleMask` exits. But there's no guarantee that the action wouldn't block or loop infinitely, so that seems pretty dicey.

Reply:
I think you're on the right track here, but it'd probably be good to add a note to the public `Exporter` interface cautioning about these instances.